Implements retryable execution feature

joker1007 opened the pull request:
Some insert errors are retryable.
For example, I use MySQL and multi threaded embulk-output-mysql, deadlock error occurs often.
If plugin can retry transaction, plugin can continue to process.

I think that such situation occurs only when `mode` is `merge_direct` or `merge` (maybe only MySQL).
But I am in trouble.

hito4t commented:
Thank you for the PR!

Can you show the stack trace when a deadlock occurred?
I'd like to know the SQL which causes a deadlock.

In regard to the source, Embulk already has a class for retry.
https://github.com/embulk/embulk/blob/master/embulk-core/src/main/java/org/embulk/spi/util/RetryExecutor.java
Can you use the class?

joker1007 commented:
Thanks, I did not know RetryExecutor.
I try to refactor this code by it.

And stack trace is here.

```
org.embulk.exec.PartialExecutionException: java.lang.RuntimeException: java.sql.BatchUpdateException: Deadlock found when trying to get lock; try restarting transaction
	at org.embulk.exec.BulkLoader$LoaderState.buildPartialExecuteException(org/embulk/exec/BulkLoader.java:363)
	at org.embulk.exec.BulkLoader.doRun(org/embulk/exec/BulkLoader.java:572)
	at org.embulk.exec.BulkLoader.access$000(org/embulk/exec/BulkLoader.java:33)
	at org.embulk.exec.BulkLoader$1.run(org/embulk/exec/BulkLoader.java:374)
	at org.embulk.exec.BulkLoader$1.run(org/embulk/exec/BulkLoader.java:370)
	at org.embulk.spi.Exec.doWith(org/embulk/spi/Exec.java:25)
	at org.embulk.exec.BulkLoader.run(org/embulk/exec/BulkLoader.java:370)
	at org.embulk.EmbulkEmbed.run(org/embulk/EmbulkEmbed.java:180)
	at java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:606)
	at RUBY.run(uri:classloader:/embulk/runner.rb:84)
	at RUBY.run(uri:classloader:/embulk/command/embulk_run.rb:306)
	at RUBY.<top>(uri:classloader:/embulk/command/embulk_main.rb:2)
	at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:937)
	at usr.bin.embulk.embulk.command.embulk_bundle.<top>(file:/usr/bin/embulk!/embulk/command/embulk_bundle.rb:30)
	at java.lang.invoke.MethodHandle.invokeWithArguments(java/lang/invoke/MethodHandle.java:599)
	at org.embulk.cli.Main.main(org/embulk/cli/Main.java:23)
Caused by: java.lang.RuntimeException: java.sql.BatchUpdateException: Deadlock found when trying to get lock; try restarting transaction
	at org.embulk.output.jdbc.AbstractJdbcOutputPlugin$PluginPageOutput.finish(AbstractJdbcOutputPlugin.java:950)
	at org.embulk.spi.PageBuilder.finish(PageBuilder.java:244)
	at org.embulk.parser.jsonl.JsonlParserPlugin.run(JsonlParserPlugin.java:183)
	at org.embulk.spi.FileInputRunner.run(FileInputRunner.java:154)
	at org.embulk.spi.util.Executors.process(Executors.java:67)
	at org.embulk.spi.util.Executors.process(Executors.java:42)
	at org.embulk.exec.LocalExecutorPlugin$DirectExecutor$1.call(LocalExecutorPlugin.java:184)
	at org.embulk.exec.LocalExecutorPlugin$DirectExecutor$1.call(LocalExecutorPlugin.java:180)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.sql.BatchUpdateException: Deadlock found when trying to get lock; try restarting transaction
	at com.mysql.jdbc.StatementImpl.handleExceptionForBatch(StatementImpl.java:1340)
	at com.mysql.jdbc.PreparedStatement.executeBatchedInserts(PreparedStatement.java:1566)
	at com.mysql.jdbc.PreparedStatement.executeBatch(PreparedStatement.java:1268)
	at org.embulk.output.jdbc.StandardBatchInsert.flush(StandardBatchInsert.java:76)
	at org.embulk.output.jdbc.StandardBatchInsert.finish(StandardBatchInsert.java:89)
	at org.embulk.output.jdbc.AbstractJdbcOutputPlugin$PluginPageOutput.finish(AbstractJdbcOutputPlugin.java:948)
	... 11 more
Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLTransactionRollbackException: Deadlock found when trying to get lock; try restarting transaction
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:377)
	at com.mysql.jdbc.Util.getInstance(Util.java:360)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:985)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3887)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3823)
	at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2435)
	at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2582)
	at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2530)
	at com.mysql.jdbc.PreparedStatement.executeInternal(PreparedStatement.java:1907)
	at com.mysql.jdbc.PreparedStatement.executeUpdate(PreparedStatement.java:2141)
	at com.mysql.jdbc.PreparedStatement.executeUpdate(PreparedStatement.java:2077)
	at com.mysql.jdbc.PreparedStatement.executeUpdate(PreparedStatement.java:2062)
	at com.mysql.jdbc.PreparedStatement.executeBatchedInserts(PreparedStatement.java:1564)
	... 15 more
```

this PR don't cover the case yet...

joker1007 commented:
@hito4t 
I refactored.
- use `RetryExecutor`
- extend original Implementation that is already exists in `AbstractJdbcOutputPlugin`

Would you review this?

hito4t commented:
Thank you!

I have some comments.


We should not retry `collectMerge` because ‘doCommit‘ is already called in `withRetry`.

The `withRetry` method has been changed from `protected` to `private`, but it should be `protected` as it was for compatibility.

Conditions for retry will vary by DBMSs.
It is more flexible if we can override `isRetryableException` method as follows.
<pre>
public abstract class AbstractJdbcOutputPlugin
...
    protected boolean isRetryableException(SQLException exception)
    {
        return isRetryableException(exception.getSQLState(), exception.getErrorCode());
    }
    protected boolean isRetryableException(String sqlState, int errorCode)
    {
        return false;
    }
</pre>

We need to call `AbstractJdbcOutputPlugin#isRetryableException` from `RetryableSQLExecution#isRetryableException`.
And we need to change some methods and some inner classes to non-static.

`MySQLOutputPlugin` can override `isRetryableException(String sqlState, int errorCode)`.

joker1007 commented:
Thank you for great advice. 😄 
I fixed them.
And I have a one point.
I changed `PluginPageOutput` class to non-static, because of `withRetry` method.
But I don't know that it is suitable.
Please comment about it.

hito4t commented:
Thank you!

> I changed PluginPageOutput class to non-static, because of withRetry method.
> But I don't know that it is suitable.

I think it is suitable because only subclasses of `AbstractJdbcOutputPlugin` will use `PluginPageOutput`, and `PluginPageOutput` needs to call a method of `AbstractJdbcOutputPlugin`.

I have some more comments.
- `final`s of `con`, `task` and `schema` in `doCommit` method are not needed.
- Although we can Google 1213 and 1205 in `MySQLOutputPlugin#isRetryableException`, it is more readable if the source has comments on the meanings of the error codes.

joker1007 commented:
@hito4t 
I have forgotten to revert `final` modifier. Thanks!
And I wrote comments about error codes.

hito4t commented:
Thank you!
I've merged.
